src/edu/psu/sweng500/team8/gui/CellGUI.java

```diff
@@ -0,0 +1,513 @@
+package edu.psu.sweng500.team8.gui;
+
+import java.awt.BorderLayout;
+import java.awt.Color;
+import java.awt.Dimension;
+import java.awt.GridBagConstraints;
+import java.awt.GridBagLayout;
+import java.awt.event.FocusAdapter;
+import java.awt.event.FocusEvent;
+import java.awt.event.MouseAdapter;
+import java.awt.event.MouseEvent;
+
+import javax.swing.BorderFactory;
+import javax.swing.JLabel;
+import javax.swing.JPanel;
+import javax.swing.JTextField;
+import javax.swing.JToggleButton;
+import javax.swing.border.Border;
+import javax.swing.border.LineBorder;
+import javax.swing.text.BadLocationException;
+import javax.swing.text.DefaultHighlighter;
+import javax.swing.text.Highlighter.HighlightPainter;
+
+import edu.psu.sweng500.team8.coreDataStructures.Cell;
+import edu.psu.sweng500.team8.coreDataStructures.CellCoordinates;
+import edu.psu.sweng500.team8.coreDataStructures.Cell.ValueType;
+import edu.psu.sweng500.team8.play.GameSession;
+
+public class CellGUI extends JPanel {
+
+	private GameSession gameSession;
+	private JPanel numberInputCell;
+	private JTextField numberInputField;
+
+	private JPanel pencilMarkInputCell;
+
+	private static int NUMBER_SIZE = 16;
+	private static int CELL_SIZE = 54;
+	private static final Border SELECTED_BORDER = BorderFactory
+			.createLineBorder(Color.blue, 3);
+	private static final Border DEFAULT_BORDER = BorderFactory
+			.createLineBorder(Color.black, 1);
+	private Cell cell;
+
+	private MouseAdapter mouseAdapter;
+	
+	/* Pencil Mark Display */
+	private JPanel pencilMarkDisplayCell;
+	private JLabel[][] pencilMarkDisplay;
+
+	public CellGUI() {
+
+		setMaximumSize(new Dimension(CELL_SIZE, CELL_SIZE));
+		setMinimumSize(new Dimension(CELL_SIZE, CELL_SIZE));
+		setPreferredSize(new Dimension(CELL_SIZE, CELL_SIZE));
+
+		setBorder(new LineBorder(new Color(0, 0, 0), 0));
+		setLayout(new java.awt.GridBagLayout());
+
+		initPencilMarkDisplayCell();
+		initNumberInputCell();
+		initPencilMarkInputCell();
+	}
+
+	public void populatePencilMark(Cell cell, GameSession gameSession) {
+
+		this.numberInputCell.setVisible(false);
+		this.pencilMarkDisplayCell.setVisible(false);
+		this.pencilMarkInputCell.setVisible(true);
+
+		pencilMarkInputCell.removeAll();
+		if (cell.hasNumber()) {
+			pencilMarkInputCell.setLayout(new BorderLayout());
+			pencilMarkInputCell.setBorder(null);
+			pencilMarkInputCell.add(buildReadOnlyTextField(cell));
+		} else {
+			pencilMarkInputCell.setLayout(new GridBagLayout());
+			pencilMarkInputCell.setBorder(DEFAULT_BORDER);
+			pencilMarkInputCell.add(buildPencilMarkInputField(cell));
+		}
+	}
+
+	public void populate(Cell cell, GameSession gameSession, boolean isRefresh, FocusAdapter focusAdapter, MouseAdapter mouseAdapter) {
+
+		this.gameSession = gameSession;
+		this.cell = cell;
+		this.mouseAdapter = mouseAdapter;
+
+		/* Clear */
+		this.numberInputField.setText("");
+		this.numberInputField.addKeyListener(new CustomKeyListener(

```

Comment:
Careful here -- this is the bug I fixed in my branch. You're adding a new key listener each time populate is called. If the user plays multiple games, each text box will have a bunch of unnecessary key listeners. This is a memory leak and they can throw exceptions.